src/base/emu-i386/simx86/interp.c

```diff
@@ -98,44 +98,42 @@ static TNode *DoClose(unsigned int PC, unsigned int Interp_LONG_CS, int mode,
 		int flags)
 {
 	unsigned int P0 = InstrMeta[0].npc;
-	TNode *G, *nextG = NULL;
+	TNode *G, *nextG;
+	int linknext;
 
-	assert(CurrIMeta >= 0);
-	/* If the code doesn't terminate with a jump/loop instruction
-	 * it still lacks the tail code; add it here */
-	IMeta *GL = &InstrMeta[CurrIMeta];
-	if (GL->gen[GL->ngen-1].op < JMP_TAILCODE) {
-		/* almost always we get here because e_querymark() in
-		   interp_post() found code. If there is no overlap
-		   and the node is compatible we can generate a jump
-		   to it and link as well */
-		nextG = FindTree(PC);
-		if (nextG && nextG->mode == mode &&
-		    nextG->cs == Interp_LONG_CS)
-			JMPGen(JMP_LINK, mode, PC);
-		else
-			JMPGen(JMP_TAILCODE, mode, PC);
-	}
-
-	assert(PC > P0);
+	assert(CurrIMeta >= 0 && PC > P0);
 	if (e_querymark(P0, PC - P0)) {
+	    /* with SPEC_PREJIT give up if we'd overwrite
+	       existing code */
+	    if (flags & F_SPRJ) {
+		CurrIMeta = -1;
+		return NULL;
+	    }
 	    InvalidateNodeRange(P0, PC - P0, NULL);
 	    if (!e_querymprotrange_full(P0, PC - P0)) {
 		unsigned int abeg, aend;
 
 		/* in case of prejit, the page content could
 		 * be altered, so we better re-jit */
-		if (flags & (F_PREJ | F_SPRJ))
+		if (flags & F_PREJ) {
+			CurrIMeta = -1;
 			return NULL;
+		}
 		abeg = P0 & _PAGE_MASK;
 		aend = PC & _PAGE_MASK;
 		/* re-populate cache */
 		for (; abeg <= aend; abeg += PAGE_SIZE)
 			Fetch(abeg);
 	    }
 	}
+
+	IMeta *GL = &InstrMeta[CurrIMeta];
+	linknext = GL->gen[GL->ngen-1].op == JMP_LINK && !(flags & F_SPRJ);
 	G = Close(PC, Interp_LONG_CS, mode, flags);
-	if (nextG)
+	if (linknext &&
+	    (nextG = FindTree(PC)) &&
```

Comment:
So you seem to add a
couple of FindTrees()'s,
and I am not even sure
its always at an instruction
boundary, or also at the
middle.
Is this good to do so?

Reply:
This is basically early linking moved down from line 112, ie if the last op is jmp_link and there is code at PC we can link to it. FindTree returns NULL if it's not the beginning of a block.

Reply:
Maybe FindTree() returns
NULL, but excessive FindTree()
calls should definitely be avoided
for performance reasons.
So is it good to add 2?